src/lib.rs

```diff
@@ -0,0 +1,198 @@
+extern crate regex;
+use regex::Regex;
+use std::error;
+use std::fmt;
+
+
+pub struct Siso {
+    route: Route,
+}
+
+impl Siso {
+    pub fn register(&mut self, path_nodes : &[PathNode], value: RouteValue) -> Result<bool, SisoError> {
+        if 0 == path_nodes.len() {
+            panic!("Need a least one node for registering a value...");
+        }
+        return self.route.register(path_nodes, value);
+    }
+
+    pub fn find(&self, path_nodes : &[PathNode]) -> Result<RouteValue, SisoError> {
+        if 0 == path_nodes.len() {
+            panic!("Need a least one node for finding a value...");
+        }
+        return self.route.find(path_nodes);
+    }
+
+    pub fn new() -> Siso {
+        Siso { route: Route::new(PathNode::PathPart(String::from("[root]")), None) }
+    }
+}
+
+#[derive(Debug)]
+struct Route {
+    node: PathNode,
+    routes: Vec<Route>,
+    value: RouteValue,
+}
+
+impl Route {
+    fn register(&mut self, path_nodes : &[PathNode], value: RouteValue) -> Result<bool, SisoError> {
+        let path_node = &path_nodes[0];
+        let mut index = Err("");
+
+        for i in 0..self.routes.len() {
+            if *path_node == self.routes[i].node {
+                index = Ok(i)
+            }
+        }
+
+        let index = match index {
+            Ok(i) => i,
+            _ => {
+                let new_route = Route::new(path_node.clone(), None);
+                self.routes.push(new_route);
+                self.routes.len() - 1
+            }
+
+        };
+
+        if path_nodes.len() == 1 {
```

Comment:
First, returns can be lifted out of `if`: `if cond { return foo() } else { return bar() }` -> `return if cond { foo() } else { bar() }`. Then, `return` and `;` could be removed as it is the last expr in the block.